### PR TITLE
ian-chiu-proposal-submission

### PR DESCRIPTION
**Pull request by ian-chiu**
---
A good PR should include the following parts:

- [X] A clear title
- [X] A meaningful message for PR, as well as its commits
- [X] from your specific branch (***not master***) merging to teacher's `HW` branch (***not master***, again)
- [X] excluding any irrelevant files, such as binaries, text files, or dot files
- [ ] passing tests/CI

### CODE REVIEW

File: project/ian-chiu/README.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+#### Neighbor
+
+There are two kinds of neighbor, Von Neumann(VN) and Moore(M). Von Neumann is like a rubik cube, the cell at the center has 26 neighbors. Moore has only 4 neighbors, top, down, front and back.
+
+### Optimization
+
+#### Multithreading and SIMD... or any parallel computing techinques
+
+Use parallel computing techinques, like multithreading and SIMD operations, to increase computation speed of calculating all cell states in the next frame under a rule.
```
Comment (yungyuc) [inline project/ian-chiu/README.md:44]: I am not familiar with the automata algorithm.  If it is not parallel friendly, you might spend a lot of time debugging the parallel algorithm.

Before a parallel algorithm can be rigorously developed, a sequential implementation is usually needed.

File: project/ian-chiu/README.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+#### Greedy Meshing Algorithm
+
+The idea of the algorithm is combining adjacent blocks to reduce the part count while visually keeping everything the same, which is useful to reduce draw calls. 
+
+For a 100x100x100 grid, we have to draw maximum 1000000 cubes if all cells are not empty. If the number of draw calls is too large, the communication between cpu and gpu will be overwhelming, which can greatly reduce the performance of rendering.
+
+![](https://i.imgur.com/SCIFdNj.png)
+
+## Prospective Users
```
Comment (yungyuc) [inline project/ian-chiu/README.md:56]: It's a comprehensive survey.  I like it.

File: project/ian-chiu/README.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+* CameraController:
+    * update: update camera view and projection matrix according to the control input
+* Application:
+    * run: control the render loop and the all update for the next frame
+    * randomize: randomize all cells in a grid
+    * clear: empty all cells in a grid
+
+### C++
+
+* calculateNextState: calculate all cells' states of the next frame. Should use multithreading and SIMD or any parallel computing techniques to increase computation speed.
```
Comment (yungyuc) [inline project/ian-chiu/README.md:116]: If multi-threading is implemented, it is correct to put it in the C++ level.

File: project/ian-chiu/README.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+5. Basic camera control: Orbiting and Zooming
+
+### Input / Output
+
+* Input: GUI, keyboard, mouse
+* Ouput: application window
+
+### Third Party Library
+
+- raylib-py or pygame: A simple video games programming library for dealing with windows, keyboard/mouse input, graphics and GUI.
```
Comment (yungyuc) [inline project/ian-chiu/README.md:98]: It makes sense to use a simple framework for GUI.

File: project/ian-chiu/README.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+## API Description
+
+### Python
+
+* Rule: 
+    * apply: apply the cell rule
```
Comment (yungyuc) [inline project/ian-chiu/README.md:105]: Would you need to call back to Python for enforcing the rules, or the rules are only data to be set in the C++ engine?

File: project/ian-chiu/README.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+
+### Week 3 (11/14):
+
+- Create the GUI.
+- Implement all the simulation operations and settings that needs GUI.
+
+### Week 4 (11/21):
+
+- Optimize computing speed of calculateNextCells by using the parallel computing techniques, like SIMD or multi-threading.
```
Comment (yungyuc) [inline project/ian-chiu/README.md:157]: I would be satisfied even if you simply optimize the code without parallelism.  Parallelism is good to have.

**Review by yungyuc — COMMENTED** [review]
LGTM.  I can merge the proposal if you want me to.

File: project/ian-chiu/README.md (hunk 1 above)
Comment (ian-chiu) [reply project/ian-chiu/README.md:44]: Yeah. Thanks for the advice. A sequential implementation will be developed first.

File: project/ian-chiu/README.md (hunk 2 above)
Comment (ian-chiu) [reply project/ian-chiu/README.md:56]: Thanks.

File: project/ian-chiu/README.md (hunk 3 above)
Comment (ian-chiu) [reply project/ian-chiu/README.md:116]: Right, I think so too.

File: project/ian-chiu/README.md (hunk 4 above)
Comment (ian-chiu) [reply project/ian-chiu/README.md:98]: Sure.

File: project/ian-chiu/README.md (hunk 6 above)
Comment (ian-chiu) [reply project/ian-chiu/README.md:157]: Ok, good to know.

File: project/ian-chiu/README.md (hunk 5 above)
Comment (ian-chiu) [reply project/ian-chiu/README.md:105]: The rule should be written in python, and C++ will get the rule from python to calculate the next state of all cells. However, I am not 100% sure. I will make the decision during development.

**Comment by ian-chiu** [comment]
> LGTM. I can merge the proposal if you want me to.

Yeah. Please merge the proposal. Thanks.

File: project/ian-chiu/README.md (hunk 5 above)
Comment (yungyuc) [reply project/ian-chiu/README.md:105]: Sounds good.

**Pull request merged**